### DISCUSSION TITLE
replace `buildTarget/dependencySources` with `workspace/libraries`

### DISCUSSION THREAD

**Discussion by abrams27** [build-server-protocol/build-server-protocol#632] [Proposals]
**"The Problem":**
so currently in order to add a new external dependency to IDEA (if you want to jar srcjar dependency) have to:
- call `buildTarget/dependencySources` (to collect src jars)
- call `buildTarget/javacOptions` (to collect jars)
- combine them somehow using a name-based heuristic

it's kinda complicated especially when build tool (at least bazel) initially provides us a structured info about the dependencies where jars and src jars are provided together

also it uses more memory, since you need to provide the same jars for multiple targets instead of providing them only once

**"Solution":**
`workspace/libraries` endpoint, which returns a structured dependencies (so src jars together with jars) and transitive dependencies of a dependency; then normal build target just depend on them like on any other build target

pros:
- more memory efficient
- easier to implement (at least for bazel & intellij)
- faster (less serialization / deserialization, parsing ...)

cons:
- im not sure how well it will scale for non-jvm langs
- how to handle generated sources? where to pass them (we just create another lib `<name of the target>_gen`

what do you think? how well can it work with other build tools? we have an "internal" implementation for our bazel-bsp and intellij-bsp which works really well for us (especially on big big projects)

our impl: 
https://github.com/JetBrains/bazel-bsp/blob/master/protocol/src/main/kotlin/org/jetbrains/bsp/WorkspaceLibraries.kt
https://github.com/JetBrains/bazel-bsp/blob/master/protocol/src/main/kotlin/org/jetbrains/bsp/BazelBuildServer.kt

cc @tpasternak

**Comment by tgodzik** [comment]
Isn't that what https://build-server-protocol.github.io/docs/specification#buildtargetdependencymodules-request does actually?

**Reply by abrams27** [reply]
possibly, but I dont see any jars / srcjars there at all, `export type DependencyModuleData = any;` doesnt seem like a nice thing to maintain : /

**Reply by tgodzik** [reply]
If we decide to go with something more structured we should probably deprecate dependency-modules I think.

**Reply by abrams27** [reply]
or we can adjust them, but there is also one big difference - `buildTarget/dependencyModules` is again called per target, we want to make it "global" - so just a one call per workspace instead of multiple calls per build target

**Reply by lefou** [reply]
The BSP client is still required to support dependencies per build target, so you don't save anything. Having dependencies defined globally is just a special case. Your server can return the same dependencies for each build target, which should not result in much overhead. The BSP client is also free to aggregate some or all build targets into one request, so that the overhead is some more bytes send over the protocol. In contrst, forcing the server and the client to negotiate whether they want to use global or local dependencies or even both, and handle potential inconsistencies between them, is much more complex than it is worth. IMHO.

**Reply by lefou** [reply]
Having some structure in the dependencies is a good thing though.

**Reply by abrams27** [reply]
> Your server can return the same dependencies for each build target, which should not result in much overhead.

we are talking about really big projects, and on top of that there has to be a mechanizm which matches the sources with jars, 

both ends we have structured deps, but we need to get rid of the structure for bsp in the middle, i wouldnt say that it's more flexible, but imo (as developer of both ends of the protocol) more chaotic tbh

> The BSP client is also free to aggregate some or all build targets into one request,

we actually have such mechanizm and it did help a bit, but the migration to this endpoint had waay bigger impact on the performance 


> Having dependencies defined globally is just a special case.

not really imo, it works the other way around as well - worst case scenario you can make libs like `<target name>-<name of the library>` and only then it would use the same amount of resources as currently


on top of that it feels more "natural" structured deps, because usually it's the case in build tools as well (e.g. maven deps)

**Reply by agluszak** [reply]
> worst case scenario you can make libs like <target name>-<name of the library> and only then it would use the same amount of resources as currently

not sure what you mean by that

**Reply by abrams27** [reply]
i edited the response, in general it's still possible to create a "target-specific" libraries with this mechanizm

**Reply by lefou** [reply]
> we are talking about really big projects,

Indeed, larger project often can take a long time to import in a BSP client (also depending on what the client is requesting as part of the import.) Maybe we need a generic way to select a subset of build targets in the BSP client. Even some kind of preset/config comes to mind. An option to opt-out after import might still take a very long time. Or a BSP client could interactively ask it's user which build targets (the first 20...) to build as part of the import.

**Reply by abrams27** [reply]
yup, but that's another issue (https://github.com/build-server-protocol/build-server-protocol/discussions/504) we have such mechanizm as well, but anyway users want to import as big as possible subset, since it just makes the development nicer

**Comment by adpi2** [comment]
Thanks for opening this discussion. I also think that the dependencies/libraries request could be more structured and compact. But I have a few remarks concerning the current proposal.

## Language agnosticity

We can't use `jars` and `srcJars` for non-JVM languages. So in the end we really need an abstract `DependencyData` that would contain language-specific information (jars and source jars for JVM, modules for JS, crates for Rust etc).

## Conflict resolution

As far as I know, I don't think that Bazel does conflict resolution but most JVM build tools do: sbt, Mill, Maven, Gradle.

Suppose I have target `a` and target `b` that depends on `a`. `a` depending on `lib:1.0.0` does not imply that `b` transitively depend on `lib:1.0.0`. Because of conflict resolution it could depend on `lib:1.1.0` instead.

## workspace vs build target

Having a `buildTarget/....` request is more flexible because the client can decide to import some build targets and not all of them.
The build target result items don't necessarily need to be independent. For instance we could have:

```ts
export interface BuildTargetDependencyResult {
  items: BuildTargetDependencyItem[];
  allDependencies: Dependency[];
}

export interface BuildTargetDependencyItem {
  target: BuildTargetIdentifier

  // all the resolved dependencies and transitive dependencies
  // because of conflict resolution
  dependencies: DependencyIdentifier[]
}

export interface Dependency {
  id: DepencenyIdentifier
  dataKind: String
  data: Any // jar and source jar
}
```
